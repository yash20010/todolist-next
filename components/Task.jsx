import styles from '../styles/Home.module.css'
import axios from 'axios'

const url = 'https://todolist-next-eta.vercel.app/api/task'

const Task = ({ task, setTask, tasks, setTasks }) => {
  const updateTask = async (id) => {
    try {
      const originalTasks = [...tasks]
      const index = originalTasks.findIndex((t) => t._id === id)
      const { data } = await axios.put(url + '/' + id, {
        completed: !originalTasks[index].completed,
      })
      originalTasks[index] = data.data
      setTasks(originalTasks)
      console.log(data.message)
    } catch (error) {
      console.log(error)
    }
  }

  const editTask = (id) => {
    const currentTask = tasks.filter((task) => task._id === id)
    setTask(currentTask[0])
  }

  const deleteTask = async (id) => {
    try {
      const { data } = await axios.delete(url + '/' + id)
      setTasks((prev) => prev.filter((task) => task._id !== id))
      console.log(data.message)
      setTask({ task: '' })
    } catch (error) {
      console.log(error)
    }
  }

  return (
    <div className={styles.task_container} data-testid="task-container">
      <input
        type="checkbox"
        className={styles.check_box}
        checked={task.completed}
        onChange={() => updateTask(task._id)}
        data-testid="checkbox"
      />
      <p
        className={
          task.completed
            ? styles.task_text + ' ' + styles.line_through
            : styles.task_text
        }
        data-testid={'task_text' + task._id}
      >
        {task.task}
      </p>
      <button
        onClick={() => editTask(task._id)}
        className={styles.edit_task}
        data-testid="edit-btn"
      >
        &#9998;
      </button>
      <button
        onClick={() => deleteTask(task._id)}
        className={styles.remove_task}
        data-testid="delete-btn"
      >
        &#10006;
      </button>
    </div>
  )
}

export default Task
