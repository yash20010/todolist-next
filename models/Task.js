import mongoose from 'mongoose'

const TaskSchema = new mongoose.Schema({
  task: { type: String, required: true },
  completed: { type: Boolean, default: false },
  createdAt: { type: Date, default: Date.now },
})

export default mongoose.models.Task || mongoose.model('Task', TaskSchema)
