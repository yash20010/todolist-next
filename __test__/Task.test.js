import React from 'react'
import ReactDOM from 'react-dom'
import Task from '../components/Task.jsx'
import { render, fireEvent, cleanup } from '@testing-library/react'
import '@testing-library/jest-dom/extend-expect'
import renderer from 'react-test-renderer'

afterEach(cleanup)

it('renders without crashing', () => {
  const div = document.createElement('div')
  ReactDOM.render(
    <Task
      task={{
        _id: 1234,
        task: 'hello',
        completed: false,
        createdAt: new Date(),
      }}
    />,
    div
  )
})

it('should render the checkbox, todo text, edit and delete button', () => {
  const { getByTestId } = render(
    <Task
      task={{
        _id: 1234,
        task: 'hello',
        completed: false,
        createdAt: new Date(),
      }}
    />
  )
  expect(getByTestId('checkbox')).toBeInTheDocument()
  expect(getByTestId('task_text1234')).toBeInTheDocument()
  expect(getByTestId('edit-btn')).toBeInTheDocument()
  expect(getByTestId('delete-btn')).toBeInTheDocument()
})

it('matches snapshot 1 - completed false', () => {
  const tree = renderer
    .create(
      <Task
        task={{
          _id: 1234,
          task: 'hello',
          completed: false,
          createdAt: new Date(),
        }}
      />
    )
    .toJSON()
  expect(tree).toMatchSnapshot()
})

it('matches snapshot 2 - completed true', () => {
  const tree = renderer
    .create(
      <Task
        task={{
          _id: 1234,
          task: 'hello',
          completed: true,
          createdAt: new Date(),
        }}
      />
    )
    .toJSON()
  expect(tree).toMatchSnapshot()
})
