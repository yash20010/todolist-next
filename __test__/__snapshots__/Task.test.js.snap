// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1 - completed false 1`] = `
<div
  className="task_container"
  data-testid="task-container"
>
  <input
    checked={false}
    className="check_box"
    data-testid="checkbox"
    onChange={[Function]}
    type="checkbox"
  />
  <p
    className="task_text"
    data-testid="task_text1234"
  >
    hello
  </p>
  <button
    className="edit_task"
    data-testid="edit-btn"
    onClick={[Function]}
  >
    ✎
  </button>
  <button
    className="remove_task"
    data-testid="delete-btn"
    onClick={[Function]}
  >
    ✖
  </button>
</div>
`;

exports[`matches snapshot 2 - completed true 1`] = `
<div
  className="task_container"
  data-testid="task-container"
>
  <input
    checked={true}
    className="check_box"
    data-testid="checkbox"
    onChange={[Function]}
    type="checkbox"
  />
  <p
    className="task_text line_through"
    data-testid="task_text1234"
  >
    hello
  </p>
  <button
    className="edit_task"
    data-testid="edit-btn"
    onClick={[Function]}
  >
    ✎
  </button>
  <button
    className="remove_task"
    data-testid="delete-btn"
    onClick={[Function]}
  >
    ✖
  </button>
</div>
`;
